[FIX JENKINS-41696] Fix recording fingerprints

foxtrot9 opened the pull request:
1. Fix bug which records fingerprints while
   ignoring the excludes when archiving artifacts.

### Proposed changelog entries

* Entry 1, bug: Fix bug which records fingerprints ignoring the excludes while archiving artifacts.
* Entry 2, rfe: Support setting excludes and case sensitivity in the `fingerprint()` build step in Pipeline and other job types
** Documentation: https://jenkins.io/doc/pipeline/steps/core/#fingerprint-record-fingerprints-of-files-to-track-usage


### Submitter checklist

- [x] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jglick @daniel-beck @oleg-nenashev

foxtrot9 committed:
[FIX JENKINS-41696] Fix recording fingerprints

1. Fix bug which records fingerprints while
   ignoring the excludes when archiving artifacts.

daniel-beck commented on core/src/main/java/hudson/tasks/ArtifactArchiver.java:
```diff
@@ -234,7 +234,8 @@ public void perform(Run<?,?> build, FilePath ws, Launcher launcher, TaskListener
                 if (fingerprint) {
-                    new Fingerprinter(artifacts).perform(build, ws, launcher, listener);
+                    new Fingerprinter(String.join(",", files.values())).perform(build, ws, launcher, listener);
+                    //new Fingerprinter(artifacts).perform(build, ws, launcher, listener);
```
```suggestion
```

foxtrot9 commented:
This bug is because of the fact that Fingerprinter.FindRecords.invoke() method includes files which should have been excluded. Currently, I have hot-fixed it in ArtifactArchiver class.
However, if maintainers feel that we should refactor Fingerprinter.FindRecords class then I am happy to do it.

daniel-beck commented:
> However, if maintainers feel that we should refactor Fingerprinter.FindRecords class then I am happy to do it.

I agree it's probably cleaner to add similar exclusion options to `Fingerprinter` so it can do the job correctly itself. Don't need to be exposed to users necessarily.

foxtrot9 commented:
Working on cleaner solution. Don't merge.

foxtrot9 committed:
[JENKINS-41696] Fix recording fingerprints

1. Fix bug which records fingerprints while
   ignoring the excludes when archiving artifacts.
2. Fingerprinter now supports excluding patterns for targets,
   caseSensitivity and defaultExcludes.

foxtrot9 commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -92,13 +94,36 @@
         this.targets = targets;
     }
 
+    // @DataBoundConstructor
```
Need feedback on how this annotation is used. I can't figure out what it does.

jglick replied:
There is already a `@DataBoundConstructor`; you may not have two.

You _may_ delete the new constructor and instead define three new `@DataBoundSetter`s. In that case you _may_ also extend `Fingerprinter/config.jelly` to allow these things to be configured from the standalone fingerprint step (not merely from its indirect usage in archive artifacts).

foxtrot9 committed:
[FIX JENKINS-41696] Fix recording fingerprints

1. Fix bug which records fingerprints while
   ignoring the excludes when archiving artifacts.
2. Fingerprinter now supports excludes,
   defaultExcludes and caseSensitivity.

Signed-off-by: Mit Naria <mit4dev@gmail.com>

jglick commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -92,13 +95,40 @@
+    /**
+     * Indicate whether include and exclude patterns should be considered as case sensitive
+     */
+    private boolean caseSensitive = true;
```
You need to be careful with field initializers which do not match the Java defaults. For Pipeline this is no problem because every `fingerprint` step call will call the `@DataBoundConstructor` and so these initializer expressions will be applied. But for steps in freestyle (etc.) projects, where the `Fingerprinter` object was serialized to `config.xml`, after an upgrade the initializer expressions will be ignored for existing steps—so while `excludes` will be `null` as expected, `defaultExcludes` and `caseSensitive` will both be `false`.

This can be addressed with a `readResolve` if you switch the latter two fields to be of type `Boolean` so you can distinguish unconfigured from old installation (`null`) from deliberately off (`false` ~ `FALSE`).

(You can also invert the sense of the fields while retaining the sense of the getters and setters using `!`, though this is arguably more confusing than using `Boolean`.)

foxtrot9 replied:
FWIU, `Boolean` implements `Serializable` interface which in turn solves our serialization issue. Am I right, here?

daniel-beck replied:
That's unrelated, Xstream just serializes everything.

The problem is that if you introduce a new field, and an object gets deserialized from a previous Jenkins version that did not have the field, it's `null` or the default value for a primitive. No constructors get called and nothing. So this will start out being `false` when someone updated their Jenkins.

jvz replied:
Will Xstream at least try to use a registered Converter? Otherwise, how would you migrate String to Secret without additional support code?

foxtrot9 replied:
@jglick Thanks for explanation.
@daniel-beck Thanks for clearing out doubt.

foxtrot9 committed:
[FIX JENKINS-41696] Fix recording fingerprints

    1. Fix bug which records fingerprints while
       ignoring the excludes when archiving artifacts.
    2. Fingerprinter now supports excludes,
       defaultExcludes and caseSensitivity.

foxtrot9 committed:
removing unused annotation and triggering a build.

foxtrot9 commented:
I believe that build is failing due to unrelated reason. I am removing WIP label and request maintainers to merge if changes are good.

varyvol commented:
@daniel-beck @jvz @jglick @oleg-nenashev could you please re-review this? Thanks!

jvz commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -92,13 +93,40 @@
+    /**
+     * Default ant exclusion
+     */
+    private Boolean defaultExcludes = true;
```
```suggestion
    private boolean defaultExcludes = true;
```

Otherwise you'll have an NPE in `getDefaultExcludes()`

jvz commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -92,13 +93,40 @@
+    /**
+     * Indicate whether include and exclude patterns should be considered as case sensitive
+     */
+    private Boolean caseSensitive = true;
```
```suggestion
    private boolean caseSensitive = true;
```

jvz commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -109,6 +137,18 @@ public String getTargets() {
+        return excludes;
+    }
+
+    public boolean getDefaultExcludes() {
```
```suggestion
    public boolean isDefaultExcludes() {
```

jvz replied:
```suggestion
    public Boolean getDefaultExcludes() {
```

jvz commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -109,6 +137,18 @@ public String getTargets() {
+        return defaultExcludes;
+    }
+
+    public boolean getCaseSensitive() {
```
```suggestion
    public boolean isCaseSensitive() {
```

jvz replied:
```suggestion
    public Boolean getCaseSensitive() {
```

jvz reviewed (commented):
I have some small code suggestions. Otherwise, this PR looks good to me.

jvz commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -92,13 +93,40 @@
+        this.excludes = excludes;
+    }
+
+    @DataBoundSetter public void setDefaultExcludes(boolean defaultExcludes) {
```
```suggestion
    @DataBoundSetter public void setDefaultExcludes(Boolean defaultExcludes) {
```

I'm only considering this one for consistency with my suggestions for the getters, though it may not be necessary.

foxtrot9 replied:
Reverted to `boolean`.

jvz commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -92,13 +93,40 @@
+        this.defaultExcludes = defaultExcludes;
+    }
+
+    @DataBoundSetter public void setCaseSensitive(boolean caseSensitive) {
```
```suggestion
    @DataBoundSetter public void setCaseSensitive(Boolean caseSensitive) {
```

Same as above.

foxtrot9 replied:
Reverted to `boolean`.

jvz reviewed (commented):
Fixed my suggestions due to misreading it originally.

jvz committed:
Apply suggestions from code review by jvz

Co-Authored-By: foxtrot9 <mit4dev@gmail.com>

foxtrot9 commented:
No problem @jvz . Commited changes.

jglick commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -92,13 +93,40 @@
+        this.excludes = excludes;
+    }
+
+    @DataBoundSetter public void setDefaultExcludes(Boolean defaultExcludes) {
```
Getters and setters should be of type `boolean` not `Boolean`.

Please reread https://github.com/jenkinsci/jenkins/pull/3915#discussion_r262178204. This is, if anything, worse since now after deserializing an old `config.xml` your code

```java
new FindRecords(targets, excludes, defaultExcludes, caseSensitive, build.getTimeInMillis())
```

will throw a `NullPointerException` unboxing e.g. a null `caseSensitive`.

jglick replied:
By the way [this](https://jenkins.io/doc/developer/plugin-development/pipeline-integration/#handling-default-values) may be helpful.

foxtrot9 replied:
@jglick You have mentioned that we can use `readResolve` method to mitigate issue of passing null Boolean object.
- I found a readResolve method in ArtiactArchiver.java file. (Path: hudson/tasks/ArtifactArchiver.java)
```
// Backwards compatibility for older builds
    @SuppressFBWarnings(value = "RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE",
            justification = "Null checks in readResolve are valid since we deserialize and upgrade objects")
    protected Object readResolve() {
        if (allowEmptyArchive == null) {
            this.allowEmptyArchive = SystemProperties.getBoolean(ArtifactArchiver.class.getName()+".warnOnEmpty");
        }
        if (defaultExcludes == null){
            defaultExcludes = true;
        }
        if (caseSensitive == null) {
            caseSensitive = true;
        }
        return this;
    }
```
Based on it, I guess that following code can be written. But I don't understand how and when readResolve method gets called.
```
// Backwards compatibility for older builds
    @SuppressFBWarnings(value = "RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE",
            justification = "Null checks in readResolve are valid since we deserialize and upgrade objects")
    protected Object readResolve() {
        if (defaultExcludes == null){
            defaultExcludes = true;
        }
        if (caseSensitive == null) {
            caseSensitive = true;
        }
        return this;
    }
```

jglick replied:
Yes, you can `readResolve` if that is the most convenient thing. It is called during XStream deserialization, in this case of a `config.xml` for a freestyle-ish project which has been configured via the GUI in a previous Jenkins session. (Pipeline projects would be creating the object via its `@DataBoundConstructor` and `@DataBoundSetter`.) It is best to read the Java serialization guide to understand what methods like this do.

A somewhat less magical alternative is to leave the field `null` and then interpret a default value in the getter and make sure all other accesses go through the getter. You still need to understand serialization: for a freestyle project, the constructor will be called only when the project is (re-)configured, not every time Jenkins starts, and this means for example that an instance initializer

```java
private Boolean caseSensitive = true;
```

will never have been called for a publisher which had been configured prior to the Jenkins upgrade represented by this patch.

If in doubt, you can always write functional tests using `@LocalData` which reproduce this kind of scenario.

foxtrot9 replied:
@jglick I added readResolve method below. I have tagged you there. You can also check my last commit-7.

foxtrot9 commented:
Pleawse ignore above commit. I will add commit for suggestions by @jglick .

daniel-beck reviewed (dismissed):
As Jesse wrote, public API needs to be `boolean` rather than `Boolean`.

foxtrot9 commented:
I am busy till end of week. I will pick it up on this weekend.

foxtrot9 committed:
[JENKINS-41696] Fix deserialization bug using readResolve.

Signed-off-by: Mit Naria <mit4dev@gmail.com>

foxtrot9 commented:
@jglick Requesting review.

daniel-beck commented:
@foxtrot9 Did you forget to push changes? I don't see what you did to respond to code review comments.

foxtrot9 commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -109,6 +137,32 @@ public String getTargets() {
+     * We ensure that fields are initialized to
+     * default values after deserialization.
+     */
+    protected Object readResolve() {
```
@daniel-beck Added readResolve method here.

foxtrot9 replied:
@jglick

foxtrot9 commented on core/src/main/resources/hudson/tasks/Fingerprinter/config.properties:
```diff
@@ -1,17 +1,17 @@
-# Copyright (c) 2004-2009, Sun Microsystems, Inc., Kohsuke Kawaguchi
-# 
+#
+# Copyright (c) 2019, Mit Naria <mit4dev@gmail.com>
```
Removing copy-pasted copyright from other files. Same goes for other license files.

foxtrot9 commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -109,6 +137,32 @@ public String getTargets() {
+     * We ensure that fields are initialized to
+     * default values after deserialization.
+     */
+    protected Object readResolve() {
```
@daniel-beck @jglick I added readResolve method here.

foxtrot9 commented on core/src/main/resources/hudson/tasks/Fingerprinter/config.properties:
```diff
@@ -0,0 +1,24 @@
+# The MIT License
+#
+# Copyright (c) 2019, Mit Naria <mit4dev@gmail.com>
```
Removing copy-pasted copyright from other file. Same goes for other license files.

jglick commented:
As in https://github.com/jenkinsci/jenkins/pull/3915#discussion_r271540578 getters and setters should not use `java.lang.Boolean`.

foxtrot9 committed:
Update Fingerprinter.java

1. Getters and setters for defaultExcludes and caseSensitive use boolean instead of Boolean.

foxtrot9 commented:
@jglick Updated. Thank you for pointing it out again.

foxtrot9 commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -109,6 +137,32 @@ public String getTargets() {
+        return excludes;
+    }
+
+    public boolean getDefaultExcludes() {
```
Reverted to `boolean`.

foxtrot9 commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -109,6 +137,32 @@ public String getTargets() {
+        return defaultExcludes;
+    }
+
+    public boolean getCaseSensitive() {
```
Reverted to `boolean`.

oleg-nenashev reviewed (approved):
I am a bit late to the party, but the current code version looks good to me. Note that the pull request is not backportable as is, because it introduces new features in the `fingerprint()` step. I have updated the changelog for it. Taking the current state and the incoming .3 merge window closure, maybe it is ok to NOT backport the PR and to just wait till the new LTS baseline in 6 weeks

fcojfernandez reviewed (approved):
LGTM

oleg-nenashev commented:
On hold for now to give other reviewers time to process feedback

jglick commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -109,6 +137,32 @@ public String getTargets() {
+     * We ensure that fields are initialized to
+     * default values after deserialization.
+     */
+    protected Object readResolve() {
```
Better `private`.

jglick commented on core/src/main/java/hudson/tasks/Fingerprinter.java:
```diff
@@ -92,13 +93,40 @@
     }
 
+    @DataBoundSetter public void setExcludes(String excludes) {
+        this.excludes = excludes;
```
Should use `Util.fixEmpty`. Otherwise there will be a mismatch between `""` as saved by the GUI form and `null` as returned by a default instance. This will result in a gratuitous

```groovy
excludes: ''
```

added in the **Snippet Generator** for Pipeline. For freestyle projects, will result in a gratuitous

```xml
<excludes></excludes>
```

jglick reviewed (changes requested):
Minor bug remaining, but the rest looks good.

foxtrot9 committed:
Minor bug fixs.

Signed-off-by: Mit Naria <mit4dev@gmail.com>

foxtrot9 commented:
Addressed comments made by @jglick .

foxtrot9 committed:
Fix stupid mistake in last commit

Signed-off-by: Mit Naria <mit4dev@gmail.com>

foxtrot9 commented:
Addressed comments made by @jglick .

foxtrot9 commented:
Addressed comments made by @jglick .

jglick reviewed (approved):
Looks right now. Thanks for your patience.

oleg-nenashev commented:
Will merge tomorrow if no negative feedback

oleg-nenashev commented:
Thanks @foxtrot9 ! It will land in the next release
